File: azkar/apple/Core/Clock.swift
// Clock times (minutes after midnight) and the daily time windows made from them.
import Foundation

/// "HH:mm" (or "H:mm") -> minutes after midnight.
func parseClock(_ s: String) -> Int? {
  let parts = s.split(separator: ":", omittingEmptySubsequences: false)
  guard parts.count == 2, parts[1].count == 2,
    let h = Int(parts[0]), let m = Int(parts[1]),
    (0..<24).contains(h), (0..<60).contains(m)
  else { return nil }
  return h * 60 + m
}

/// Minutes after midnight -> "HH:mm".
func formatClock(_ minutes: Int) -> String {
  String(format: "%02d:%02d", minutes / 60, minutes % 60)
}

func minuteOfDay(_ date: Date, _ calendar: Calendar) -> Int {
  let c = calendar.dateComponents([.hour, .minute], from: date)
  return c.hour! * 60 + c.minute!
}

func dayKey(_ date: Date, _ calendar: Calendar) -> String {
  let c = calendar.dateComponents([.year, .month, .day], from: date)
  return String(format: "%04d-%02d-%02d", c.year!, c.month!, c.day!)
}

/// [start, end) in minutes after midnight; wraps past midnight when end < start.
struct TimeWindow: Equatable {
  var start: Int
  var end: Int

  func contains(minuteOfDay m: Int) -> Bool {
    start <= end ? (start <= m && m < end) : (m >= start || m < end)
  }
}

File: azkar/kotlin/shared/src/commonMain/kotlin/com/wildduck/azkar/core/Clock.kt
// Clock times (minutes after midnight) and the daily time windows made from them.
package com.wildduck.azkar.core

import kotlinx.datetime.LocalDateTime

/** "HH:mm" (or "H:mm") -> minutes after midnight. */
fun parseClock(s: String): Int? {
    val parts = s.split(":")
    if (parts.size != 2 || parts[1].length != 2) return null
    val h = parts[0].toIntOrNull() ?: return null
    val m = parts[1].toIntOrNull() ?: return null
    if (h !in 0..23 || m !in 0..59) return null
    return h * 60 + m
}

/** Minutes after midnight -> "HH:mm". */
fun formatClock(minutes: Int): String =
    (minutes / 60).toString().padStart(2, '0') + ":" + (minutes % 60).toString().padStart(2, '0')

fun minuteOfDay(time: LocalDateTime): Int = time.hour * 60 + time.minute

/** "yyyy-MM-dd": the day that today's morning/evening progress belongs to. */
fun dayKey(time: LocalDateTime): String = time.date.toString()

/** [start, end) in minutes after midnight; wraps past midnight when end < start. */
data class TimeWindow(val start: Int, val end: Int) {
    operator fun contains(minuteOfDay: Int): Boolean =
        if (start <= end) minuteOfDay >= start && minuteOfDay < end else minuteOfDay >= start || minuteOfDay < end
}
